src/runtime/openglcompute.cpp

```diff
@@ -339,6 +340,29 @@ WEAK int halide_openglcompute_device_free(void *user_context, halide_buffer_t *b
     return 0;
 }
 
+namespace {
+
+template <typename Source, typename Dest>
+__attribute__((always_inline)) void converting_copy_memory_helper(const device_copy &copy, int d, int64_t src_off, int64_t dst_off) {
+    // Skip size-1 dimensions
+    while (d >= 0 && copy.extent[d] == 1) d--;
+
+    if (d == -1) {
+        const Source *from = (Source *)(copy.src + src_off);
+        Dest *to = (Dest *)(copy.dst + dst_off);
+        for (uint64_t index = 0; index < copy.chunk_size; index++) {
+            *to++ = (Dest)*from++;
+        }
+    } else {
+        for (uint64_t i = 0; i < copy.extent[d]; i++) {
+            converting_copy_memory_helper<Dest, Source>(copy, d - 1, src_off, dst_off);
```

Comment:
(1) Order of template arguments is reversed in this call vs. declaration order above. Intentional?
(2) Making a recursive call to an always_inline function? (Not a bug, I guess, just another litany of C++ wtf)